packages/core/database/src/utils/identifiers/index.ts

```diff
@@ -21,25 +41,164 @@ type NameOptions = {
   maxLength?: number;
 };
 
-export const getNameFromTokens = (tokens: NameToken[], max: number = MAX_DB_IDENTIFIER_LENGTH) => {
-  const fullLength = tokens
+// returns a hash of length len
+export function createHash(data: string, len: number): string {
+  if (!_.isInteger(len) || len <= 0) {
+    throw new Error(`createHash length must be a positive integer, received ${len}`);
+  }
+
+  // TODO: shake256 is based on SHA-3 and is slow, we don't care about cryptographically secure, only uniqueness and speed
+  //       investigate alternatives before releasing this. But it is only run on startup, so it should be fine.
+  const hash = crypto.createHash('shake256', { outputLength: Math.ceil(len / 2) }).update(data);
+  return hash.digest('hex').substring(0, len);
+}
+
+export function tokenWithHash(name: string, len: number) {
+  if (!_.isInteger(len) || len <= 0) {
+    throw new Error(`tokenWithHash length must be a positive integer, received ${len}`);
+  }
+  if (name.length <= len) {
+    return name;
+  }
+  if (len < MIN_TOKEN_LENGTH + HASH_LENGTH) {
+    throw new Error(
+      `length for part of identifier too short, minimum is hash length (${HASH_LENGTH}) plus min token length (${MIN_TOKEN_LENGTH}), received ${len} for token ${name}`
```

Comment:
```suggestion
      `length is too short, the minimum is ${HASH_LENGTH + MIN_TOKEN_LENGTH}, received ${len} for token ${name}`
```

I'm not sure what value the longer message adds to a user who doesn't really need to understand the implementation detail of how it's calculate?

Reply:
Yeah, I had the same though, but it should never be possible for the user to see unless we screw up pretty badly (like combining 5 compressible tokens plus a suffix, for example)

And then even if we do, this could be caught at the layer that actually calls this method (the schema creation) and convert it to a more user-friendly message